File: javascript/problems/155-minStack.js
/**
 * initialize your data structure here.
 */
var MinStack = function() {
    this.vals = [];
    this.minVals = [];
};

/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
    if (!this.minVals.length || x <= this.getMin()) this.minVals.push(x);
    this.vals.push(x);
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    if (this.top() === this.getMin()) this.minVals.pop();
    this.vals.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.vals[this.vals.length - 1];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.minVals[this.minVals.length - 1];
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: csharp/Solution/155-MinStack.cs
using System.Collections.Generic;

namespace Solution
{
    public class MinStack
    {
        private readonly LinkedList<int> _vals;
        private readonly LinkedList<int> _minVals;

        public MinStack()
        {
            _vals = new LinkedList<int>();
            _minVals = new LinkedList<int>();
        }

        public void Push(int x)
        {
            if (_minVals.Count == 0 || x <= _minVals.Last.Value) _minVals.AddLast(x);
            _vals.AddLast(x);
        }

        public void Pop()
        {
            if (Top() == GetMin()) _minVals.RemoveLast();
            _vals.RemoveLast();
        }

        public int Top()
        {
            return _vals.Last.Value;
        }

        public int GetMin()
        {
            return _minVals.Last.Value;
        }
    }
}
